File: Java/src/main/java/com/DSA/Java/Algorithms/Sorting/MergeSort.java
package com.DSA.Java.Algorithms.Sorting;

public class MergeSort {
   int arr[];
   private void merge(int[] arr, int left, int middle, int right) {
      int[] tempLeftArray = new int[middle - left + 2];
      int[] tempRightArray = new int[right - middle + 1];

      for (int i = 0; i <= middle - left; i++) {
         tempLeftArray[i] = arr[left + i];
      }
      for (int i = 0; i < right - middle; i++) {
         tempRightArray[i] = arr[middle + i + 1];
      }
      
      tempLeftArray[middle - left + 1] = Integer.MAX_VALUE;
      tempRightArray[right - middle] = Integer.MAX_VALUE;

      int i = 0, j = 0;
      for (int k = left; k <= right; k++) {
         if (tempLeftArray[i] < tempRightArray[j]) {
            arr[k] = tempLeftArray[i];
            i++;
         } else {
            arr[k] = tempRightArray[j];
            j++;
         }
      }
   }

   private void mergeSort(int[] arr, int left, int right) {
      if (right > left) {
         int middle = (left + right) / 2;
         mergeSort(arr, left, middle);
         mergeSort(arr, middle + 1, right);
         merge(arr, left, middle, right);
      }
   }

   public int[] sort(int[] arr) {
      this.arr = arr;
      mergeSort(arr, 0, arr.length - 1);
      return arr;
   }
}

File: C++/Algorithm/src/Sorting/MergeSort.cpp
#include "../../include/Sorting/MergeSort.h"

void merge_sort_helper(std::vector<int> &arr, int left, int right, int middle) {
    int temp_left_array[middle - left + 2];
    int temp_right_array[right - middle + 1];

    for (int i = 0; i <= middle - left; i++) {
        temp_left_array[i] = arr[left + i];
    }

    for (int i = 0; i < right - middle; i++) {
        temp_right_array[i] = arr[middle + i + 1];
    }
    
    temp_left_array[middle - left + 1] = INT_MAX;
    temp_right_array[right - middle] = INT_MAX;

    int i = 0, j = 0;
    for (int k = left; k <= right; k++) {
        if (temp_left_array[i] < temp_right_array[j]) {
            arr[k] = temp_left_array[i];
            i++;
        } else {
            arr[k] = temp_right_array[j];
            j++;
        }
    }
}

void merge_sort_helper_2(std::vector<int> &arr, int left, int right) {
    if (right > left) {
        int middle = (right + left) / 2;
        merge_sort_helper_2(arr, left, middle);
        merge_sort_helper_2(arr, middle + 1, right);
        merge_sort_helper(arr, left, right, middle);
    }
}

void MergeSort::sort_array() {
    merge_sort_helper_2(arr, 0, arr.size() - 1);
}
